test/docker-e2e/dockerchain/testchain.go

```diff
@@ -0,0 +1,119 @@
+package dockerchain
+
+import (
+	"context"
+	"github.com/celestiaorg/celestia-app/v4/app"
+	"github.com/celestiaorg/celestia-app/v4/app/encoding"
+	"github.com/celestiaorg/celestia-app/v4/pkg/user"
+	"github.com/celestiaorg/celestia-app/v4/test/util/genesis"
+	tastoradockertypes "github.com/celestiaorg/tastora/framework/docker"
+	"github.com/celestiaorg/tastora/framework/testutil/config"
+	"github.com/celestiaorg/tastora/framework/testutil/maps"
+	cometcfg "github.com/cometbft/cometbft/config"
+	cmtjson "github.com/cometbft/cometbft/libs/json"
+	"github.com/cometbft/cometbft/privval"
+	servercfg "github.com/cosmos/cosmos-sdk/server/config"
+	"github.com/cosmos/cosmos-sdk/types/module/testutil"
+	"github.com/stretchr/testify/require"
+	"testing"
+	"time"
+)
+
+// NewCelestiaChainBuilder constructs a new ChainBuilder configured for a Celestia instance with predefined parameters.
+func NewCelestiaChainBuilder(t *testing.T, cfg *Config) *tastoradockertypes.ChainBuilder {
+	genesisBz, err := cfg.Genesis.ExportBytes()
+	require.NoError(t, err, "failed to export genesis bytes")
+
+	// TODO: why do I need to do this?
+	genesisBz, err = maps.SetField(genesisBz, "consensus", map[string]interface{}{})
+	require.NoError(t, err)
+	genesisBz, err = maps.SetField(genesisBz, "consensus.params.version.app", "4")
+	require.NoError(t, err)
+
+	encodingConfig := testutil.MakeTestEncodingConfig(app.ModuleEncodingRegisters...)
+
+	kr := cfg.Genesis.Keyring()
+
+	// TODO: accountNames can't be hard coded
+	accountNames := []string{"validator", "val1", "val2"}
+	vals := make([]tastoradockertypes.ChainNodeConfig, 3)
+	for i := 0; i < len(accountNames); i++ {
+		privKeyBz := getValidatorPrivateKeyBytes(t, cfg.Genesis, i)
+		vals[i] = tastoradockertypes.NewChainNodeConfigBuilder().
+			WithPrivValidatorKey(privKeyBz).
+			WithAccountName(accountNames[i]).
+			WithKeyring(kr).
+			Build()
+	}
+
+	return tastoradockertypes.NewChainBuilder(t).
+		WithName("celestia"). // just influences home directory on the host.
+		WithChainID(cfg.Genesis.ChainID).
+		WithDockerClient(cfg.DockerClient).
+		WithDockerNetworkID(cfg.DockerNetworkID).
+		WithImage(tastoradockertypes.NewDockerImage(cfg.Image, cfg.Tag, "10001:10001")).
+		WithAdditionalStartArgs("--force-no-bbr", "--grpc.enable", "--grpc.address", "0.0.0.0:9090", "--rpc.grpc_laddr=tcp://0.0.0.0:9099").
+		WithEncodingConfig(&encodingConfig).
+		WithPostInit(getPostInitModifications("0.025utia")...).
```

Comment:
post init functions let us modify anything on the node filesystem before start is called, so basically allows full control over what all the config files look like.